Write support for tables implemented from extensions

alessandrogario opened the pull request:
### Description
This PR adds support for writable extension tables by providing an xUpdate sqlite3 implementation.

### Disclamer
Write support for core/built-in tables is **NOT** implemented, and will not work.

### Demo
This is a small demo for the following example: *osquery/examples/example_writable_table.cpp*
```
osquery> select * from WritableTable;
osquery> insert into WritableTable (text, integer) values ("1", 1);
osquery> insert into WritableTable (text, integer) values ("1", 1);
Error: constraint failed
osquery> insert into WritableTable (text, integer) values ("2", 2);
osquery> select * from WritableTable;
+------+---------+
| text | integer |
+------+---------+
| 2    | 2       |
| 1    | 1       |
+------+---------+
osquery> delete from WritableTable where text = "1";
osquery> select * from WritableTable;
+------+---------+
| text | integer |
+------+---------+
| 2    | 2       |
+------+---------+
osquery> insert into WritableTable (text, integer) values ("3", 3);
osquery> select * from WritableTable;
+------+---------+
| text | integer |
+------+---------+
| 3    | 3       |
| 2    | 2       |
+------+---------+
osquery> update WritableTable set text="2", integer=2 where text="3";
Error: constraint failed
osquery> update WritableTable set text="5", integer=5 where text="3";
osquery> select * from WritableTable;
+------+---------+
| text | integer |
+------+---------+
| 5    | 5       |
| 2    | 2       |
+------+---------+
```

obelisk commented:
This is really interesting, what potential use cases do you envision for this sort of thing?

obelisk commented on osquery/sql/virtual_table.cpp:
```diff
@@ -169,25 +419,29 @@ int xCreate(sqlite3* db,
   auto status =
       Registry::call("table", name, {{"action", "columns"}}, response);
   if (!status.ok() || response.size() == 0) {
-    delete pVtab->content;
```
Why are these being removed? (I'm probably just missing something)

obelisk commented on osquery/sql/virtual_table.cpp:
```diff
@@ -169,25 +419,29 @@ int xCreate(sqlite3* db,
                << "): " << getStringForSQLiteReturnCode(rc);
+
     VLOG(1) << "Cannot create virtual table using: " << statement;
-    delete pVtab->content;
```
As above.

obelisk commented on osquery/sql/virtual_table.cpp:
```diff
@@ -582,59 +843,83 @@ static int xFilter(sqlite3_vtab_cursor* pVtabCursor,
+  sqlite3_module* module =
+      tables::sqlite::getVirtualTableModule(name, is_extension);
+  if (module == nullptr) {
+    auto error_message =
```
I would just do this in the VLOG(1). In my opinion, the only time that returning an error message this verbose in a Status is useful is if you are checking it in code, or are going to print it out. Since you've already printed it out, you probably don't need to return it and `Status(1)` should suffice.

cc: @muffins Do you agree with this as a design decision? Please tell me I'm wrong if you think so.

obelisk commented on include/osquery/tables.h:
```diff
@@ -752,7 +777,7 @@ class TablePlugin : public Plugin {
  protected:
   /// An SQL table containing the table definition/syntax.
-  std::string columnDefinition() const;
+  std::string columnDefinition(bool is_extension) const;
```
This change could break other people who are compiling extensions against this header correct? Could we provide default values instead?

alessandrogario replied:
The columnDefinition method was defined several times (as a method and as a function) both overloaded and with default parameters, and this made it hard to read because you never knew where the call was going when passing some parameters (and this is why I didn't provide a default value here).

Isn't this method only used by osquery and not by the extension developers? (I'm asking because I didn't have to change the other example extensions after this patch).

obelisk replied:
It is possible that other people are using them because they are in our public headers. This (and the changes I marked API Change) could in theory break anyone who is using the SDK. If this is the only way to do it, or this is a great improvement in the API then we can merge this but we shouldn't do so without evaluating the possibility of breaking some third party extensions.

obelisk commented on include/osquery/tables.h:
```diff
@@ -846,10 +871,6 @@ class TablePlugin : public Plugin {
                                     PluginRequest& request);
 
-  /// Helper data structure transformation methods.
-  static void setContextFromRequest(const PluginRequest& request,
```
API Change

alessandrogario replied:
I moved it because it wasn't referenced anywhere! Thanks for spotting this!

obelisk commented on include/osquery/tables.h:
```diff
@@ -869,18 +890,20 @@ class TablePlugin : public Plugin {
 
 /// Helper method to generate the virtual table CREATE statement.
-std::string columnDefinition(const TableColumns& columns);
+std::string columnDefinition(const TableColumns& columns, bool is_extension);
```
API Change.

obelisk commented on include/osquery/tables.h:
```diff
@@ -869,18 +890,20 @@ class TablePlugin : public Plugin {
 /// Helper method to generate the virtual table CREATE statement.
 std::string columnDefinition(const PluginResponse& response,
-                             bool aliases = false);
+                             bool aliases,
```
API Change.

alessandrogario commented on osquery/examples/example_writable_table.cpp:
```diff
@@ -0,0 +1,236 @@
+  /// Callback for DELETE queries
+  QueryData delete_(QueryContext& context, const PluginRequest& request) {
+    auto rowid = request.at("id");
+    if (id_to_internal_id.count(rowid) == 0) {
```
Yeah, you are right! I'm not sure why but I ended up matching some other code I found in the other files without thinking! Going to fix it right now

obelisk commented on osquery/examples/example_writable_table.cpp:
```diff
@@ -0,0 +1,231 @@
+    row_data.rowid = new_rowid;
+
+    // Build the primary key and make sure it's unique
+    std::string internal_rowid =
```
This should be abstracted as a function that generates the primary key of the table. Currently adding a new column to the primary key is going to require changes to places and will be error prone (even if you use all the right columns but in the wrong order).

obelisk commented on osquery/examples/example_writable_table.cpp:
```diff
@@ -0,0 +1,231 @@
+    }
+
+    // Build the primary key and make sure it's unique
+    std::string new_internal_id =
```
Call new primary key function.

obelisk commented on osquery/examples/example_writable_table.cpp:
```diff
@@ -0,0 +1,231 @@
+
+    auto current_internal_id = internal_id_it->second;
+
+    // This variable contains a value for each column in the table, separated by
```
Is this still true?

obelisk commented on osquery/examples/example_writable_table.cpp:
```diff
@@ -0,0 +1,231 @@
+    auto new_rowid_it = request.find("new_id");
+
+    std::string new_rowid;
+    if (new_rowid_it != request.end()) {
```
This could be a ternary.

```
std::string new_rowid = new_rowid_it != request.end() ? new_rowid_it->second : current_rowid;
```

obelisk commented on osquery/examples/example_writable_table.cpp:
```diff
@@ -0,0 +1,231 @@
+
+  /// Callback for INSERT queries
+  QueryData insert(QueryContext& context, const PluginRequest& request) {
+    // This variable contains a value for each column in the table, separated by
```
Is this still true?

obelisk commented on osquery/examples/example_writable_table.cpp:
```diff
@@ -0,0 +1,231 @@
+    QueryData results;
+
+    for (const auto& row_descriptor : row_map) {
+      const auto row_data = row_descriptor.second;
```
I think this can be a reference

obelisk commented on osquery/core/tables.cpp:
```diff
@@ -91,35 +101,42 @@ void TablePlugin::setContextFromRequest(const PluginRequest& request,
+    response = update(context, request);
+
+  } else if (action == "columns") {
+    static_cast<void>(request);
```
What does this do?

alessandrogario replied:
This is just to make static analysis (and warnings with -Wunused) happy; I kind of keep adding it due to habits for other projects! I can see it's not applicable here, I'll remove it!

obelisk commented on osquery/examples/example_writable_table.cpp:
```diff
@@ -0,0 +1,222 @@
+class WritableTable : public TablePlugin {
+ private:
+  /// Contents for a single row
+  struct RowData final {
```
Can you elaborate on why you chose to use a `struct` here instead of an `unordered_map`? It seems to me that if these were maps you could just return them in your gen function instead of copying the values to the `Row` object. Maybe you can just use `Row` objects as well? Another upside is that most of your int conversions will go away.

The only draw back I see is not being able to store types in the their native types (they all become strings).

Let me know if I missed something obvious about your implementation :p

alessandrogario replied:
I couldn't use strings because the data is now serialized using rapidjson, automatically matching the type of each column

obelisk commented on osquery/examples/example_writable_table.cpp:
```diff
@@ -0,0 +1,222 @@
+  }
+
+  /// Expands a value list returned by osquery into a RowData structure
+  bool getRowData(RowData& row_data, const std::string& value_list) const {
```
This should return a status

obelisk commented on osquery/examples/example_writable_table.cpp:
```diff
@@ -0,0 +1,222 @@
+    rapidjson::Document document;
+    document.Parse(value_list);
+    if (document.HasParseError() || !document.IsArray()) {
+      VLOG(1) << "Invalid format";
```
Put these into the message of the status

obelisk commented on osquery/examples/example_writable_table.cpp:
```diff
@@ -0,0 +1,222 @@
+    }
+
+    if (document.Size() != 2U) {
+      VLOG(1) << "Wrong column count.";
```
As above.

obelisk commented on osquery/examples/example_writable_table.cpp:
```diff
@@ -0,0 +1,222 @@
+      row_data.text = document[0].GetString();
+    }
+
+    if (document[0].IsNull()) {
```
Nit. Could be ternary.

obelisk commented:
ok to test

osqueryer commented:
:-1: The commit d10a5ee546f1bf48e00589a03f46f1ee524942c1 (Job results: [3349](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3349/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit d10a5ee546f1bf48e00589a03f46f1ee524942c1 (Job results: [3350](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3350/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit b4eb5187a0ac717f087ae96688bbc63a6d44d2f5 (Job results: [3352](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3352/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit b4eb5187a0ac717f087ae96688bbc63a6d44d2f5 (Job results: [3353](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3353/)) failed one or more tests (**Windows**).

fmanco commented:
Hi @alessandrogario, why is it you say "Write support for core/built-in tables is NOT implemented, and will not work."?

obelisk commented:
@fmanco Building in tables won't work because the code is generated for those tables early in the build process without the insert method set. We would have to change a lot more code and create some more context to support this in core tables.

alessandrogario commented on osquery/sql/virtual_table.cpp:
```diff
@@ -73,6 +72,132 @@ static inline std::string opString(unsigned char op) {
+std::vector<std::string> extension_table_list;
+std::mutex extension_table_list_mutex;
+
+bool isExtensionTable(const std::string& table_name) {
```
I tried to use the registry many times, but it doesn't look like it can be used reliably while it is initializing the tables

alessandrogario commented on osquery/sql/virtual_table.cpp:
```diff
@@ -582,59 +906,78 @@ static int xFilter(sqlite3_vtab_cursor* pVtabCursor,
+
+struct sqlite3_module* getVirtualTableModule(const std::string& table_name,
+                                             bool extension) {
+  static std::unordered_map<std::string, struct sqlite3_module> modules;
```
This is done to actually avoid making the code messier (because we are reducing the scope as much as possible since it is not needed outside the dedicated interface function). This is also thread safe (as long as we compile with C++11 or better).

I can move this outside and remove the static if you prefer

obelisk replied:
Makes sense :)

uptycs-nishant replied:
Hi @alessandrogario I believe initialization 'static std::unordered_map<std::string, struct sqlite3_module> modules;' is thread safe but 'modules[table_name] = {};' is not thead safe
I.e. initialization of a function local static object is thread safe not the operations on them.

alessandrogario replied:
Yes! I used the wrong terms; initialization is performed on a single thread (you can see the original code doesn't protect access to this variable)

uptycs-nishant replied:
In the present code usage of variable 'module' is thread safe since -
1. It is initialized as static variable.
2. After initialization it is being used as read-only.
While the code you are proposing 'modules' variable is not read-only.

alessandrogario replied:
You are right it is not accessed in read only, but I was under the assumption that initialization was single threaded, when in fact it is not. I'll address this issue. Thanks!

uptycs-nishant replied:
"it does seem that everything is performed in a unique call from a single thread, so it doesn't matter whether you lock it or not."
I observed that the code is written with keeping thread safety in mind. Until now access to module object was being done in thread safe manner since it is being accessed as read-only. At present only single thread of control is accessing this code but it may not be the case in future.  

I believe attachlock will not be enough since it is synchronizing access to instance but modules object  can be accessed across the instances. Please correct me if I am wrong.

alessandrogario replied:
You may be right! I've ended up reverting the last commit, and adding a mutex. Thanks!

alessandrogario commented on osquery/examples/example_writable_table.cpp:
```diff
@@ -0,0 +1,299 @@
+  }
+
+  /// Saves the given row
+  Status saveRow(const Row& row, PrimaryKey primary_key = std::string()) {
```
I need to keep them separate so that the statements can be blocked earlier (i.e. without modifying the current data or calling the rowid generator) in case of a constraint failure. If I do this check inside the saveRow call, then I would have to undo changes

alessandrogario commented on osquery/examples/example_writable_table.cpp:
```diff
@@ -0,0 +1,299 @@
+
+  /// Expands a value list returned by osquery into a Row (without the rowid
+  /// column)
+  Status getRowData(Row& row, const std::string& value_list) const {
```
I kept value_list because it is more descriptive than just 'json' (it is in fact a list of values, regardless of the format). Maybe we can use json_value_array

alessandrogario commented on osquery/examples/example_writable_table.cpp:
```diff
@@ -0,0 +1,299 @@
+
+    row["text"] = document[0].IsNull() ? "" : document[0].GetString();
+
+    auto value = document[1].IsNull() ? 0 : document[1].GetInt();
```
Duh! Was trying to keep short lines :D Will fix right now!

alessandrogario commented on osquery/examples/example_writable_table.cpp:
```diff
@@ -0,0 +1,299 @@
+
+    // Finally, save the row; also pass the primary key we calculated so that
+    // the function doesn't have to compute it again
+    status = saveRow(row, primary_key);
```
See https://github.com/facebook/osquery/pull/4094#discussion_r165934266

osqueryer commented:
:-1: The commit 312e938c2b8bca34ec281fea7c035b83177b0a77 (Job results: [3358](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3358/)) failed one or more tests (**Windows**).

alessandrogario commented:
I'm not really sure why osqueryer commented that the test has failed; it is passing on my virtual machine, and it is not shown as failing in the check list at the bottom of the page!

osqueryer commented:
:-1: The commit 1f6decbc991e89b3cb4a7fa4acb6f392badf3af2 (Job results: [6528](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6528/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 6774a504e42afb0cd730964f35b27208db2e971f (Job results: [2792](https://jenkins.osquery.io/job/osqueryPullRequestOSX/2792/)) failed one or more tests (**macOS/OS X**).

osqueryer commented:
:-1: The commit 6774a504e42afb0cd730964f35b27208db2e971f (Job results: [6764](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6764/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 6774a504e42afb0cd730964f35b27208db2e971f (Job results: [1229](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/1229/)) failed one or more tests (**FreeBSD**).

osqueryer commented:
:-1: The commit 6774a504e42afb0cd730964f35b27208db2e971f (Job results: [1230](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/1230/)) failed one or more tests (**FreeBSD**).

osqueryer commented:
:-1: The commit 6774a504e42afb0cd730964f35b27208db2e971f (Job results: [6765](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6765/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit b391387bb83cd847b1362965bb1e9b0cbee796ec (Job results: [6796](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6796/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit b391387bb83cd847b1362965bb1e9b0cbee796ec (Job results: [6797](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6797/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 45b582e71e1001d6952618a88320539fd1f254dd (Job results: [3692](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3692/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit 45b582e71e1001d6952618a88320539fd1f254dd (Job results: [6798](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6798/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 45b582e71e1001d6952618a88320539fd1f254dd (Job results: [3693](https://jenkins.osquery.io/job/osqueryPullRequestWindows/3693/)) failed one or more tests (**Windows**).

osqueryer commented:
:-1: The commit 45b582e71e1001d6952618a88320539fd1f254dd (Job results: [6799](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6799/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 967a99c7997c825228ff955621753168b3a10560 (Job results: [6800](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6800/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 967a99c7997c825228ff955621753168b3a10560 (Job results: [6801](https://jenkins.osquery.io/job/osqueryPullRequestLinux/6801/)) failed one or more tests (**Linux**).

osqueryer commented:
:-1: The commit 48f5e7f771a82ccd64a429a369da9b6f41bed2fd (Job results: [1267](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/1267/)) failed one or more tests (**FreeBSD**).

osqueryer commented:
:-1: The commit 48f5e7f771a82ccd64a429a369da9b6f41bed2fd (Job results: [1268](https://jenkins.osquery.io/job/osqueryPullRequestFreeBSD/1268/)) failed one or more tests (**FreeBSD**).

theopolis commented on osquery/sql/sqlite_util.cpp:
```diff
@@ -192,12 +192,16 @@ Status SQLiteSQLPlugin::attach(const std::string& name) {
 
-  auto statement = columnDefinition(response);
+  bool is_extension = true;
+  auto statement = columnDefinition(response, false, is_extension);
```
Confirm that this `attach` method is only called for external tables.

theopolis commented on osquery/sql/virtual_table.cpp:
```diff
@@ -702,14 +1094,16 @@ void attachVirtualTables(const SQLiteDBInstanceRef& instance) {
   }
 
   PluginResponse response;
+  bool is_extension = false;
```
TODO(Ted): Double check this.

theopolis commented on osquery/core/tables.cpp:
```diff
@@ -268,6 +270,13 @@ std::string columnDefinition(const TableColumns& columns) {
+  // Tables implemented by extension can be made read/write; make sure to always
+  // keep the rowid column; this makes them easier to manage, and also removes
+  // any limit on the primary key count
+  if (is_extension) {
```
Let's see if we can fix the `INDEX=True`, so we wont need this bool.

theopolis commented on osquery/sql/virtual_table.cpp:
```diff
@@ -73,6 +72,173 @@ static inline std::string opString(unsigned char op) {
+std::mutex extension_table_list_mutex;
+
+bool isExtensionTable(const std::string& table_name) {
+  std::lock_guard<std::mutex> locker(extension_table_list_mutex);
```
You can use `ReadLock lock(...);`

theopolis commented on osquery/sql/virtual_table.cpp:
```diff
@@ -605,59 +977,79 @@ static int xFilter(sqlite3_vtab_cursor* pVtabCursor,
+
+struct sqlite3_module* getVirtualTableModule(const std::string& table_name,
+                                             bool extension) {
+  std::lock_guard<std::mutex> lock(sqlite_module_map_mutex);
```
nb: `ReadLock lock...`;

theopolis commented on osquery/sql/virtual_table.cpp:
```diff
@@ -73,6 +72,173 @@ static inline std::string opString(unsigned char op) {
+
+namespace {
+// A map containing an sqlite module object for each virtual table
+std::unordered_map<std::string, struct sqlite3_module> sqlite_module_map;
```
TODO(Ted): Review this

theopolis commented on include/osquery/tables.h:
```diff
@@ -759,6 +759,33 @@ class TablePlugin : public Plugin {
+                            const PluginRequest& request) {
+    static_cast<void>(context);
+    static_cast<void>(request);
+
```
Let's replicate the existing error messages: `Error: table osquery_info may not be modified`.

alessandrogario replied:
With the latest commit, tables can now return custom error messages. By default (i.e. when the insert/update/delete methods are not implemented) read only is returned, showing the message you mentioned

theopolis commented on osquery/sql/virtual_table.cpp:
```diff
@@ -144,8 +310,199 @@ int xNext(sqlite3_vtab_cursor* cur) {
+
+  std::string table_name = pVtab->content->name;
+
+  if (FLAGS_table_delay > 0 && pVtab->instance->tableCalled(content)) {
```
You can delete this.

theopolis commented on osquery/sql/virtual_table.cpp:
```diff
@@ -144,8 +310,199 @@ int xNext(sqlite3_vtab_cursor* cur) {
+    // Apply an optional sleep between table calls.
+    sleepFor(FLAGS_table_delay);
+  }
+  pVtab->instance->addAffectedTable(content);
```
You can delete this.

theopolis commented on osquery/sql/virtual_table.cpp:
```diff
@@ -144,8 +310,199 @@ int xNext(sqlite3_vtab_cursor* cur) {
+
+  // The SQLite instance communicates to the TablePlugin via the context.
+  QueryContext context(content);
+  context.useCache(pVtab->instance->useCache());
```
You can delete this.

theopolis commented on osquery/sql/virtual_table.cpp:
```diff
@@ -144,8 +310,199 @@ int xNext(sqlite3_vtab_cursor* cur) {
+      return serializerError;
+    }
+
+    plugin_request.insert({"json_value_array", json_value_array});
```
Investigate the use of this map key, why `json_value_array`?

theopolis commented on osquery/sql/virtual_table.cpp:
```diff
@@ -144,8 +310,199 @@ int xNext(sqlite3_vtab_cursor* cur) {
+      return serializerError;
+    }
+
+    plugin_request.insert({"json_value_array", json_value_array});
```
Again

akindyakov commented on include/osquery/tables.h:
```diff
@@ -759,6 +759,33 @@ class TablePlugin : public Plugin {
+  /// Callback for DELETE statements
+  virtual QueryData delete_(QueryContext& context,
+                            const PluginRequest& request) {
+    static_cast<void>(context);
```
May I ask you to use in such cases `boost::ignore_unused` as more explicit way to switch off this warning. [ref](https://www.boost.org/doc/libs/1_67_0/libs/core/doc/html/core/ignore_unused.html)

akindyakov commented on osquery/devtools/shell.cpp:
```diff
@@ -119,7 +119,12 @@ static char zTimerHelp[] =
+    "list",
+    "semi",
+    "csv",
+    "pretty",
```
May I ask you to submit formatting changes in separate PR?

alessandrogario replied:
This sometimes happens when we update the `clang-format ` executable from the brew repository and there are changes on how code is formatted by the tool. If I remove this, then the `make audit` step will fail (and the CI workers will fail too).

These lines will disappear once we rebase on master after the correctly formatted code is committed.

alessandrogario replied:
That's really cool! Has this been introduced recently? I had several PR fail due to this in the recent past!

akindyakov commented on osquery/devtools/shell.cpp:
```diff
@@ -1083,7 +1088,7 @@ inline void meta_schema(int nArg, char** azArg) {
               "CREATE TABLE %s%s;\n",
               table.c_str(),
-              osquery::columnDefinition(response, true).c_str());
+              osquery::columnDefinition(response, true, false).c_str());
```
I'd like to suggest you make such calls more explicit in terms of passing arguments. Smth like this:
```c++
auto const aliases = true;
auto const is_extension = false;
osquery::columnDefinition(response, aliases, is_extension).c_str());
```
What do you think?

akindyakov commented on osquery/sql/virtual_table.cpp:
```diff
@@ -74,6 +73,194 @@ static inline std::string opString(unsigned char op) {
+  return (std::find(extension_table_list.begin(),
+                    extension_table_list.end(),
+                    table_name) != extension_table_list.end());
+}
```
0. There are 3 things that are used only together, would you like to make a make a class for them?
1. There is no usage of it outside of cpp file. Would you like to put it under anonymous namespace? 
2. As far as order of elements in `extension_table_list` is not important (I didn't find it, may be I'm wrong) I'd suggest to use `unordered_set` instead of it.

akindyakov commented on osquery/sql/virtual_table.cpp:
```diff
@@ -74,6 +73,194 @@ static inline std::string opString(unsigned char op) {
+
+bool getColumnValue(std::string& value,
+                    int index,
+                    int argc,
```
`index` and `argc` are used for indexing, would you like to define them as `std::size_t`? To avoid a problems with negative values.

alessandrogario replied:
See https://github.com/facebook/osquery/pull/4094#discussion_r202101631

akindyakov commented on osquery/sql/virtual_table.cpp:
```diff
@@ -74,6 +73,194 @@ static inline std::string opString(unsigned char op) {
+  }
+
+  default: {
+    VLOG(1) << "Invalid column type";
```
Should it be a `LOG(ERROR)` ?
Would you like to provide more context in the error message?

akindyakov commented on osquery/sql/virtual_table.cpp:
```diff
@@ -74,6 +73,194 @@ static inline std::string opString(unsigned char op) {
+
+    auto buffer_size = sqlite3_value_bytes(sqlite_value);
+    value.resize(buffer_size);
+    std::memcpy(&value[0], data_pointer, buffer_size);
```
You could do the same operation without using low level function `memcpy`.
Smth like this:
```c++
value.assign(data_pointer, buffer_size);
```

akindyakov commented on osquery/sql/virtual_table.cpp:
```diff
@@ -74,6 +73,194 @@ static inline std::string opString(unsigned char op) {
+  }
+
+  case SQLITE_NULL: {
+    break;
```
Should we put in the `value` something like "0" or similar in this case?

alessandrogario replied:
The `value` variable is cleared before we enter the switch, so in case of a NULL we get an empty string. This is how NULL values are generally handled by tables

akindyakov commented on osquery/sql/virtual_table.cpp:
```diff
@@ -74,6 +73,194 @@ static inline std::string opString(unsigned char op) {
+
+// Type-aware serializer to pass parameters between osquery and the extension
+int serializeUpdateParameters(std::string& json_value_array,
+                              int argc,
```
Would you like to consider of using `std::size_t` for `argc` variable?
And also define it as `const`?

alessandrogario replied:
See https://github.com/facebook/osquery/pull/4094#discussion_r202101631

akindyakov commented on osquery/sql/virtual_table.cpp:
```diff
@@ -74,6 +73,194 @@ static inline std::string opString(unsigned char op) {
+                              int argc,
+                              sqlite3_value** argv,
+                              const TableColumns& columnDescriptors) {
+  if (columnDescriptors.size() != static_cast<size_t>(argc - 2)) {
```
It could be problematic to substitute something from number and cast it to `unsigned` type. Is here any possibility for `argc` to be less than `2`?

alessandrogario replied:
All the argc parameters are defined as `int` to match the type used by sqlite3 to define that variable. 

If we switch the type to something else, we will have to cast it to `size_t` when we receive it and back to `int` when we return it back to the library (would you prefer this approach?).

Parameter serialization only happens when an UPDATE or INSERT statement has been received.

This is a short explanation of how it works for this case, but the documentation goes in more detail:

### Update queries
We are updating an existing row; the ID for that row is inside argv[0]

##### argv[1] == NULL
The table code is expected to keep using the ID associated with the row being modified. This ID can be found in argv[0].

##### argv[1] == INTEGER
The table handler is expected to update the row, invalidate the existing row ID (argv[0]) and use the new one found in argv[1].

### Insert queries
In this case, the row is new so there is no existing row ID. This means that argv[0] will be set to SQLITE_NULL.

##### argv[1] == NULL
The table is expected to come up with a new row ID and return it back to sqlite.

##### argv[1] == INTEGER
sqlite requires the table logic to use the supplied row ID when saving the row.

akindyakov replied:
Thank you for the explanation. I see your point. But even so I'd like to suggest you to use a `std::size_t` here.
1. `int` in `sqlite3` interface is just a part of interface. And is good to see this conversion from `std::size_t` to `int` only when we use the `sqlite` library.
2. Mixing up a signed and unsigned types is bad idea. It could be cause for many bugs and even UB, which is pure evil. Look at the example:
```c++
#include <iostream>
int main() {
  const auto sv = -2;
  const auto usv = 1u;
  auto res = sv + usv;
  std::cout << res << '\n';
  return 0;
}
```
What is the type of `res` and what is the answer? This is potential UB.
3. You right for high load pieces of code static conversion from one type to another could be have a notable performance penalty. But we can not solve this problem by using a `int` type here. Because there still will be a lot of conversions. And, that is the worst, they will be implicit. Which is hard to find, to control, test and debug.
4. And semantics, `std::size_t` is a special type dedicated to represent indexes, address offsets and so on. If you see that type you will know - that variable is for the size. `int` is too common, variable of this type could be anything.

akindyakov commented on osquery/sql/virtual_table.cpp:
```diff
@@ -74,6 +73,194 @@ static inline std::string opString(unsigned char op) {
+      break;
+    }
+
+    default: { return SQLITE_MISMATCH; }
```
Should we write something to `LOG(ERROR)` here?

muffins commented:
@alessandrogario looks like @theopolis and @akindyakov had a few more things to fix, and this needs a rebase, but we should be getting close on this yeah?

alessandrogario commented:
Hey @muffins!

I've rebased and fixed the issues that have been mentioned; the remaining comments are personal notes/bookmarks that Ted has left for his review during QueryCon. I don't think he is requesting for any change, but correct me if I'm wrong!
